pricelist_item_generator/models/pricelist.py

```diff
@@ -0,0 +1,415 @@
+            default['product_condition_ids'] = self._get_o2m_values(
+                self.product_condition_ids, 'product_condition_id')
+        return super(PricelistItemGenerator, self).copy(default)
+
+    @api.multi
+    def _get_o2m_values(self, objects, field_name):
+        """ get values from One2many"""
+        self.ensure_one()
+        values = self._get_vals_from_objs(objects)
+        values = [{y: z for y, z in dic.items() if y != field_name}
+                  for dic in values]
+        return [(0, 0, x) for x in values]
+
+    @api.multi
+    def write(self, vals):
+        if 'name' in vals or 'sequence' in vals:
+            vals['to_update'] = True
+        return super(PricelistItemGenerator, self).write(vals)
+
+    @api.multi
+    def build_pricelist_item(self):
+        """ Called by the UI
+        """
+        prd_list_item_m = self.env['product.pricelist.item']
+        for gen in self:
+            tmpl_objs = defaultdict(list)
+            condition_objs = {}
+            # Firstly we are creating missing records
+            for item in gen.item_template_ids:
+                if item.todo == 'create':
+                    tmpl_objs[item] = [x for x in gen.product_condition_ids]
+            for condition in gen.product_condition_ids:
+                if condition.todo == 'create':
+                    condition_objs[condition] = [
+                        x for x in gen.item_template_ids
+                        if x.todo != 'create']
+            for key, val in condition_objs.items():
+                for elm in val:
+                    tmpl_objs[elm].append(key)
+            for tmpl, cond_objs in tmpl_objs.items():
+                vals_conditions = gen._get_vals_from_objs(cond_objs)
+                tmpl_vals = gen._get_vals_from_objs([tmpl])
+                for vals in vals_conditions:
+                    values = gen._synchronize_items(tmpl_vals[0], vals)
+                    prd_list_item_m.create(values)
+            self._unset_todo_create(tmpl_objs, condition_objs)
+            # Then we update all values
+            price_items = prd_list_item_m.search(
+                [('price_generator_id', '=', gen.id)])
+            map_items = {'%s-%s' % (
+                         str(x.product_condition_id.id),
+                         str(x.item_template_id.id)): x.id
+                         for x in price_items}
+            tmpl_vals = gen._get_vals_from_objs(gen.item_template_ids)
+            condition_vals = gen._get_vals_from_objs(gen.product_condition_ids)
+            for tmpl in tmpl_vals:
+                for condition in condition_vals:
+                    values = gen._synchronize_items(tmpl, condition)
+                    p_list_item_id = map_items['%s-%s' % (
+                        values.get('product_condition_id'),
+                        values.get('item_template_id'))]
+                    prd_list_item_m.browse(p_list_item_id).write(values)
```

Comment:
It would be nice if this model had a M2M relation w/ `product.pricelist.item`. You could then utilize that col and the Odoo ORM to group all of these writes in with the one below, which would be significantly more efficient.